### PR TITLE
Add basic Nim support

### PR DESCRIPTION
**Pull request by barcharcraz**
This is my first pass at nim support, currently it treats nim as it's own compiler, it does not treat it as a generator for C code. I *think* this is the way to go, but I'd be interested to hear thoughts on this (esp comparing to vala support).

Currently staticlibs are broken due to an upstream bug (problems with spaces in paths....)

### CODE REVIEW

**Comment by ignatenkobrain** [comment]
Flake8 detected 16 issues on de6c87a86bd518494bf49c025f59c33e32a46435
Visit https://sider.review/gh/19784232/pull_requests/3776 to review the issues.

**Comment by jpakkane** [comment]
What packages would need to be added to the CI image so the tests can be run?

**Comment by barcharcraz** [comment]
Not sure? How can I look at the base image for CI?

both fedora and ubuntu have a "nim" package

windows binaries are available from https://nim-lang.org/install_windows.html

All that's needed is the nim compiler, a c toolchain for it to use, and meson.

**Comment by jpakkane** [comment]
The Docker file is in the `ciimage` directory. Please add the `nim` package there as part of this MR.

Due to Reasons (TM) the image is not autoupdated on merge requests, only manually. I have just uploaded a version of the image that has `nim` installed. If you have set `run_project_tests.py` up correctly, it should run the tests properly the next time this MR is pushed to.

**Comment by barcharcraz** [comment]
Ah from 0.17.2 (which is in bionic) to 0.18 (which is in fedora-28) nim started responding to --version on stdout instead of stderr.

**Comment by barcharcraz** [comment]
OK that failing test case is the spaces issue

**Comment by barcharcraz** [comment]
Upstream bug https://github.com/nim-lang/Nim/issues/8118

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -1285,6 +1288,28 @@ def generate_vala_compile(self, target, outfile):
         element.write(outfile)
         return other_src[0], other_src[1], vala_c_src
 
+    def generate_nim_target(self, target, outfile):
+        nim = target.compilers['nim']
+        relsrc = []
+        for i in target.get_sources():
+            if not nim.can_compile(i):
+                raise InvalidArguments('Rust target %s contains a non-nim source file.' % target.get_basename())
```
Comment (Akaricchi) [inline mesonbuild/backend/ninjabackend.py]: >Rust target

Looks like a copypaste error.

**Comment by jpakkane** [comment]
A test that should be added is building a mixed Nim + C (and/or C++). The tests should have both Nim having the main function calling into C and C having the main function calling into Nim. How does Nim handle that? Can you have languages of different type in a single target or do you need to

**Comment by barcharcraz** [comment]
so by default nim drives the whole thing, using C as just a intermediary language. You can specify the --compileOnly option to suppress that and have it just output C sources. The thing is the C sources are not cross platform, or even cross compiler compatible, so you need to make sure the compiler passed to nim (--cc) matches what you actually use to compile.

There's also the --genDeps command that generates a .deps file (possibly in the -MM format, but I'm not sure).

**Comment by jpakkane** [comment]
In that case we can get started by having a Nim main program that calls into C and C++ code that is compiled in the same target (or a static library). This is important to get working at the start because it requires setting up all the magic linker steps needed to combine languages.

**Comment by barcharcraz** [comment]
OK that static lib (or shlib) is the easiest, for calling code from a C TU that's part of the same target I'd like to chat about what the compilation model should actually be. There is the `--clib:` option of the compiler as well as `--passl`. `--clib` does some searching for the lib depending on the platform, `--passl` just adds stuff to the linker command line.

As I mentioned the other option is to forget about all this and just always call the nim compiler with the `--compileOnly --genDeps` options and then compile the resulting output files ourselves.

File: mesonbuild/compilers/nim.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        output_name = os.path.join(work_dir, 'nimtest')
+        code = '' # blank files are valid nim
+        args  = self.get_cross_extra_flags(environment, link=False)
+        with self.compile(code, args, 'compile') as p:
+            if p.returncode != 0:
+                msg = 'Nim compiler {!r} can not compile programs'.format(self.name_string())
+                raise EnvironmentException(msg)
+        # with open(source_name, 'w'):
+        #     pass
+
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/nim.py]: __[Flake8]__

[W293] blank line contains whitespace

File: mesonbuild/compilers/nim.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        self.language = 'nim'
+        super().__init__(exelist, version)
+        self.id = 'nim'
+        self.is_cross = False
+
+    def sanity_check(self, work_dir, environment):
+        source_name = os.path.join(work_dir, 'sanity.nim')
+        output_name = os.path.join(work_dir, 'nimtest')
+        code = '' # blank files are valid nim
+        args  = self.get_cross_extra_flags(environment, link=False)
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/nim.py]: __[Flake8]__

[E221] multiple spaces before operator

File: mesonbuild/compilers/nim.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import os.path
+import subprocess
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/nim.py]: __[Flake8]__

[F401] 'subprocess' imported but unused

File: mesonbuild/compilers/nim.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    def __init__(self, exelist, version):
+        self.language = 'nim'
+        super().__init__(exelist, version)
+        self.id = 'nim'
+        self.is_cross = False
+
+    def sanity_check(self, work_dir, environment):
+        source_name = os.path.join(work_dir, 'sanity.nim')
+        output_name = os.path.join(work_dir, 'nimtest')
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/nim.py]: __[Flake8]__

[F841] local variable 'output_name' is assigned to but never used

File: mesonbuild/compilers/nim.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+class NimCompiler(Compiler):
+
+    def __init__(self, exelist, version):
+        self.language = 'nim'
+        super().__init__(exelist, version)
+        self.id = 'nim'
+        self.is_cross = False
+
+    def sanity_check(self, work_dir, environment):
+        source_name = os.path.join(work_dir, 'sanity.nim')
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/nim.py]: __[Flake8]__

[F841] local variable 'source_name' is assigned to but never used

File: mesonbuild/compilers/nim.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        output_name = os.path.join(work_dir, 'nimtest')
+        code = '' # blank files are valid nim
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/nim.py]: __[Flake8]__

[F841] local variable 'code' is assigned to but never used

File: mesonbuild/compilers/nim.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
+import os.path
+import subprocess
+from ..mesonlib import EnvironmentException, Popen_safe, mlog
+from ..mesonlib import get_compiler_for_source
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/nim.py]: __[Flake8]__

[F401] '..mesonlib.get_compiler_for_source' imported but unused

File: mesonbuild/compilers/nim.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        args  = self.get_cross_extra_flags(environment, link=False)
+        args += self.get_output_args(output_name)
+        args += self.get_outdir_args(work_dir)
+        with open(source_name, 'w'):
+            pass
+        
+        mlog.debug('Running compile:')
+        mlog.debug('Working directory: ', work_dir)
+        mlog.debug('Command line: ', ' '.join(args), '\n')
+        pc, stdo, stde  = Popen_safe(
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/nim.py]: __[Flake8]__

[E221] multiple spaces before operator

**Comment by jpakkane** [comment]
The Travis build failure seems relevant:

    ar: 648f9a7f18/libnimlib.a: No such file or directory
    Error: execution of an external program failed: 'ar rcs /root/b 648f9a7f18/libnimlib.a  nimlib@sta/nimcache/prog.o nimlib@sta/nimcache/stdlib_system.o'

**Comment by barcharcraz** [comment]
Yep. Known issue with how nim invokes the linker (its in nim as of a few days ago. Will also be fixed by treating nim like Vala and having meson compile the generated files

**Comment by keszybz** [comment]
The nim bug with spaces is fixed. What about this PR?

**Comment by barcharcraz** [comment]
I've been working on polishing it up. There are some issues w.r.t dependency scanning. Basically everything needs to be treated as a header type dependency, annoyingly nim can't currently output a deps file while also compiling the project, and the deps file isn't in the usual format (it's one line per dep, absolute paths).

While I'm at it I was trying to have meson handle the generated C files as well, like it does with vala.

File: mesonbuild/backend/ninjabackend.py [hunk 10]
```diff
@@ -1091,6 +1095,42 @@ def generate_vala_compile(self, target, outfile):
...
+            if not nim.can_compile(i):
+                raise InvalidArguments('Nim target %s contains a non-nim source file.' % target.get_basename())
+            if main_nim_file is None:
+                main_nim_file = i.rel_to_builddir(self.build_to_src)
+            relsrc.append(i.rel_to_builddir(self.build_to_src))
+        target_name = os.path.join(target.subdir, target.get_filename())
+        # nim always outputs dep files to the name of the entry point nim file
+        main_file_name = os.path.splitext(main_nim_file.fname)[1]
+        nimcache_dir = nim.get_nimcache_dir(self.get_target_private_dir(target))
+        main_json_file = os.path.join(nimcache_dir, main_file_name + '.json')
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__

[F841] local variable 'main_json_file' is assigned to but never used

File: mesonbuild/backend/ninjabackend.py [hunk 11]
```diff
@@ -1091,6 +1095,42 @@ def generate_vala_compile(self, target, outfile):
...
+        main_deps_file = os.path.join(nimcache_dir, main_file_name + '.deps')
+        args = nim.get_always_args()
+        args += nim.get_dependency_gen_args()
+        # we need compile only with gendeps, if you try and compile
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__

[W291] trailing whitespace

File: mesonbuild/backend/ninjabackend.py [hunk 12]
```diff
@@ -16,6 +16,7 @@
 import re
 import shlex
 import pickle
+import json
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__

[F401] 'json' imported but unused

File: mesonbuild/backend/ninjabackend.py [hunk 13]
```diff
@@ -1091,6 +1095,42 @@ def generate_vala_compile(self, target, outfile):
...
+        main_json_file = os.path.join(nimcache_dir, main_file_name + '.json')
+        main_deps_file = os.path.join(nimcache_dir, main_file_name + '.deps')
+        args = nim.get_always_args()
+        args += nim.get_dependency_gen_args()
+        # we need compile only with gendeps, if you try and compile 
+        # with gendeps it either crashes ( nim <= 0.18) or does not actually
+        # compile (nim == 0.19) This comment was written when 0.19 was the latest nim version
+        args += nim.get_compile_only_args()
+        args += nim.get_buildtype_args(self.get_option_for_target('buildtype', target))
+
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__

[W293] blank line contains whitespace

File: mesonbuild/backend/ninjabackend.py [hunk 14]
```diff
@@ -1483,6 +1523,17 @@ def generate_rust_compile_rules(self, compiler, outfile, is_cross):
         outfile.write(depstyle)
         outfile.write('\n')
 
+    def generate_nim_compile_rules(self, compiler, outfile):
+        rule = 'rule %s_COMPILER\n' % compiler.get_language()
+        invoc = ' '.join([ninja_quote(i) for i in compiler.get_exelist()])
+        command = ' command = %s $ARGS $in\n' % invoc
+        depfile = ' depfile = $targetdep\n'
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__

[F841] local variable 'depfile' is assigned to but never used

**Comment by barcharcraz** [comment]
I'm closing until I have something reliably working and more ready for review.

**Comment by gavr123456789** [comment]
Any news?

**Comment by zeeev** [comment]
looking forward to nim support, bump.

**Comment by DrSensor** [comment]
??

**Comment by barcharcraz** [comment]
The main thing blocking this PR is that I don't have time and don't need the feature anymore, if someone wants to pick it up I'd be fine with that.

**Pull request closed without merging**